### PR TITLE
refactor: prevent repeating entity hydration in MtM relation

### PR DESCRIPTION
**Pull request by hustlahusky**
Prevent entity state reset caused by repeating hydration call when loading MtM relation

### CODE REVIEW

File: tests/ORM/Fixtures/RbacItemAbstract.php [hunk 1]
```diff
@@ -26,9 +31,10 @@ class RbacItemAbstract
      */
     public $children;
 
-    public function __construct(string $name)
+    public function __construct(string $name, ?string $description = null)
```
Comment (roxblnfk) [inline tests/ORM/Fixtures/RbacItemAbstract.php:34]:
```suggestion
    public function __construct(string $name, string $description = null)
```

File: tests/ORM/Fixtures/RbacItemAbstract.php [hunk 2]
```diff
@@ -14,6 +14,11 @@ class RbacItemAbstract
      */
     public $name;
 
+    /**
+     * @var string|null
+     */
```
Comment (roxblnfk) [inline tests/ORM/Fixtures/RbacItemAbstract.php:19]:
```suggestion
    /** @var string|null */
```

File: src/Iterator.php [hunk 3]
```diff
@@ -18,25 +18,38 @@
  */
 final class Iterator implements \IteratorAggregate
 {
-    /** @var ORMInterface */
+    /**
+     * @var ORMInterface
+     */
```
Comment (roxblnfk) [inline src/Iterator.php:23]:
```suggestion
    /** @var ORMInterface */
```

File: src/Iterator.php [hunk 4]
```diff
@@ -18,25 +18,38 @@
...
     private $orm;
 
-    /** @var string */
-    private $class;
+    /**
+     * @var string
+     */
```
Comment (roxblnfk) [inline src/Iterator.php:28]:
```suggestion
    /** @var string */
```

File: src/Iterator.php [hunk 5]
```diff
@@ -18,25 +18,38 @@
...
+    private $role;
 
-    /** @var iterable */
+    /**
+     * @var iterable
+     */
```
Comment (roxblnfk) [inline src/Iterator.php:33]:
```suggestion
    /** @var iterable */
```

File: src/Iterator.php [hunk 6]
```diff
@@ -18,25 +18,38 @@
...
     private $source;
 
+    /**
+     * @var bool
+     */
```
Comment (roxblnfk) [inline src/Iterator.php:38]:
```suggestion
    /** @var bool */
```

File: tests/ORM/ManyToManySingleEntityTest.php [hunk 7]
```diff
@@ -157,4 +157,39 @@ public function testClearAndFillRelation(): void
...
+        $role = new RbacRole('superAdmin', 'description');
+
+        $permission = new RbacPermission('writeUser');
+
+        $role->children->add($permission);
+        $permission->parents->add($role);
+
+        $tr = new Transaction($this->orm);
+        $tr->persist($role);
+        $tr->run();
```
Comment (roxblnfk) [inline tests/ORM/ManyToManySingleEntityTest.php:172]:
```suggestion
        $this->save($role);
```

File: tests/ORM/ManyToManySingleEntityTest.php [hunk 8]
```diff
@@ -157,4 +157,39 @@ public function testClearAndFillRelation(): void
...
+
+        // unlink
+        $fetchedRole->children->removeElement($fetchedPermission);
+        $fetchedPermission->parents->removeElement($fetchedRole);
+
+        self::assertSame('updated description', $fetchedRole->description);
+
+        $tr = new Transaction($this->orm);
+        $tr->persist($fetchedRole);
+        $tr->run();
```
Comment (roxblnfk) [inline tests/ORM/ManyToManySingleEntityTest.php:193]:
```suggestion
        $this->save($fetchedRole);
```

If here you check that no exceptions will be thrown, then write about it in the comments.
You can also check the number of read or write requests by methods:
```php
        $this->captureWriteQueries();
        $this->captureReadQueries();

        $this->save($fetchedRole);

        $this->assertNumWrites(2);
        $this->assertNumReads(0);
```

**Review by roxblnfk — COMMENTED** [review]
PR looks good.
Please do not use `Force Push` when making changes. Just add new commit.

File: tests/ORM/ManyToManySingleEntityTest.php (hunk 8 above)
Comment (hustlahusky) [reply tests/ORM/ManyToManySingleEntityTest.php]: No, I'm checking that role description will not be overwritten with current database value after loading permission parents.

Second save called for `BaseTest::assertEntitySynced()` not throw exceptions.

Comment (roxblnfk) [reply tests/ORM/ManyToManySingleEntityTest.php]: > Second save called for BaseTest::assertEntitySynced() not throw exceptions.

Then this instead:

```php
$this->orm = $this->orm->withHeap(new Heap());
```

**Review by roxblnfk — APPROVED** [review]

File: tests/ORM/ManyToManySingleEntityTest.php (hunk 8 above)
Comment (hustlahusky) [reply tests/ORM/ManyToManySingleEntityTest.php]: Thx. Done

**Pull request merged**